cstore_reader.c

```diff
@@ -147,78 +137,167 @@ CStoreBeginRead(const char *filename, TupleDesc tupleDescriptor,
+	if (footerLength + postscriptSize + CSTORE_POSTSCRIPT_SIZE_LENGTH != footerData->len)
 	{
 		ereport(ERROR, (errmsg("invalid footer size")));
 	}
 
-	footerOffset = postscriptOffset - footerLength;
-	footerBuffer = ReadFromFile(tableFooterFile, footerOffset, footerLength);
-	tableFooter = DeserializeTableFooter(footerBuffer);
+	/*
+	 * DeserializeTableFooter expects footer data without postscript, reduce len
+	 * field to trim the right end of footer data.
+	 */
+	footerData->len = footerLength;
+	tableFooter = DeserializeTableFooter(footerData);
 
-	freeResult = FreeFile(tableFooterFile);
-	if (freeResult != 0)
+	pfree(footerData->data);
+	pfree(footerData);
+
+	return tableFooter;
+}
+
+
+/*
+ * ReadFooterData reads and returns footer data. The first page contains the
+ * metadata about footer data. The function validates the metadata, reads
+ * footer data from where it is stored and returns it.
+ */
+static StringInfo
+ReadFooterData(Relation relation)
+{
+	int32 totalFooterLength = 0;
+	BlockNumber totalBlockCount = 0;
+	BlockNumber blockIndex = 0;
+	const int sizeofFooterLengthType = sizeof(int32);
+	StringInfo footerData = NULL;
+	Buffer firstBuffer = InvalidBuffer;
+	Page page = NULL;
+	char *pageData = NULL;
+	int pageDataLength = 0;
+	PageHeader pageHeader = NULL;
+	Buffer footerMetadataBuffer = InvalidBuffer;
+	Page headerPage = NULL;
+	char *headerData = NULL;
+	uint32 headerDataLength = 0;
+	uint32 startingBlock = 0;
+	uint32 blockCount = 0;
+
+	totalBlockCount = RelationGetNumberOfBlocksInFork(relation, FOOTER_FORKNUM);
+
+	if (totalBlockCount == 0)
 	{
-		ereport(ERROR, (errcode_for_file_access(),
-						errmsg("could not close file: %m")));
+		return NULL;
 	}
 
-	return tableFooter;
+	footerMetadataBuffer =  ReadBufferExtended(relation, FOOTER_FORKNUM, 0, RBM_NORMAL,
+									   NULL);
+	LockBuffer(footerMetadataBuffer, BUFFER_LOCK_SHARE);
+	headerPage = BufferGetPage(footerMetadataBuffer);
+	headerData = PageGetContents(headerPage);
+	pageHeader = (PageHeader) headerPage;
+	headerDataLength =  pageHeader->pd_lower - SizeOfPageHeaderData;
+
+	if (!DeserializeTableFooterMetadata(headerData, headerDataLength,
+										&startingBlock, &blockCount))
+	{
+		UnlockReleaseBuffer(footerMetadataBuffer);
+		return NULL;
+	}
+
+	firstBuffer = ReadBufferExtended(relation, FOOTER_FORKNUM, startingBlock,
+									 RBM_NORMAL, NULL);
+
+	LockBuffer(firstBuffer, BUFFER_LOCK_SHARE);
+	page = BufferGetPage(firstBuffer);
+	pageHeader = (PageHeader) page;
+	pageData = PageGetContents(page);
+
+	memcpy(&totalFooterLength, pageData, sizeofFooterLengthType);
+
+	if (totalFooterLength < CSTORE_POSTSCRIPT_SIZE_LENGTH)
+	{
+		ereport(ERROR, (errmsg("invalid cstore file")));
+	}
+
+	footerData = makeStringInfo();
+
+	pageDataLength = pageHeader->pd_lower - SizeOfPageHeaderData - sizeofFooterLengthType;
+
+	appendBinaryStringInfo(footerData, pageData + sizeofFooterLengthType, pageDataLength);
+
+	UnlockReleaseBuffer(firstBuffer);
+
```

Comment:
Might be worth to do a `enlargeStringInfo()` before iterating, to avoid repeated reallocations.